README.rst
Easy CLI tool for exporting and importing Nifi templates to and from XML files.

ChangeLog
CHANGES
=======

0.1.5
-----

* removed the changelog from .gitignore
* deleted unnecessary spacing in setup.cfg
* added pbr to build-requirements.txt
* fixed some issues introduced in a45c240
* updated NifiInstance methods to use the fixed api functions instead of requests
* change return values of some NifiInstance methods that didn't make sense
* refactored some coupling issues
* updated nipyapi to 0.8.0, cleaned up requirements.txt

0.1.4
-----

* fixed relative import issue2

0.1.3
-----

* another change to entry\_points..
* fixing another entry\_point issue
* fixed setup.cfg missing entry\_point section, added ChangeLog to .gitignore

0.1.0
-----

* removed redundant magics in \_\_init\_\_.py
* prepped for pypi release
* reverted init
* prepping for pbr packaging
* added docstrings, cleaned up a bit
* init commit


nifi_deploy/nifi.py
"""
Provides a NifiInstance class facilitating easy to use
methods utilizing the NiPyApi (https://github.com/Chaffelson/nipyapi)
wrapper library.
"""

import os
import sys
import getpass

from nipyapi import nifi, config, templates, canvas

# Disable urllib3 certificate warnings
from requests.packages.urllib3 import disable_warnings
disable_warnings()


class NifiInstance:
    """ The NifiInstance class facilitating easy to use
    methods utilizing the NiPyApi (https://github.com/Chaffelson/nipyapi)
    wrapper library.

    Arguments:
        url         (str): Nifi host url, defaults to environment variable `NIFI_HOST`.
        username    (str): Nifi username, defaults to environment variable `NIFI_USERNAME`.
        password    (str): Nifi password, defaults to environment variable `NIFI_PASSWORD`.
        verify_ssl  (bool): Whether to verify SSL connection - UNUSED as of now.

    """

    def __init__(self, url=None, username=None, password=None, verify_ssl=False):
        config.nifi_config.host = self._get_url(url)
        config.nifi_config.verify_ssl = verify_ssl
        config.nifi_config.username = username
        self._authenticate(username, password)


    def _get_url(self, url):
        if not url:
            try:
                url = os.environ['NIFI_HOST']
            except KeyError:
                url = input('Nifi host: ')
        if not '/nifi-api' in url:
            if not url[-1] == '/':
                url = url + '/'
            url = url + 'nifi-api'
        return url


    def _authenticate(self, username=None, password=None):
        if not username:
            try:
                config.nifi_config.username = os.environ['NIFI_USERNAME']
            except KeyError:
                config.nifi_config.username = input('Username: ')

        if not password:
            try:
                password = os.environ['NIFI_PASSWORD']
            except KeyError:
                password = getpass.getpass('Password: ')

        access_token = None
        try:
            access_token = nifi.AccessApi().create_access_token(username=config.nifi_config.username,password=password)            
        except nifi.rest.ApiException as e:
            print('Exception when calling AccessApi->create_access_token: %s\n'.format(e))

        config.nifi_config.api_key[username] = access_token
        config.nifi_config.api_client = nifi.ApiClient(header_name='Authorization', header_value='Bearer {}'.format(access_token))        


    def create_template(self, pg_id, name, desc=''):
        """ Create a template from process group id.

        Arguments:
            pg_id   (str): Process group ID to create the template from.
            name    (str): Name of the template to create.
            desc    (str): Optional, description of the template to create.
        
        Returns:
            nipyapi.nifi.TemplateEntity 
        """

        try:
            obj = templates.create_template(
                pg_id=pg_id,
                name=name,
                desc=desc
                )
        except nifi.rest.ApiException as e:
            print(e.body)
            return None

        return obj.template

    
    def delete_template(self, template_id):
        """ Delete a template from Nifi template registry.

        Arguments:
            template_id (str): ID of the template to delete.
        
        Returns:
            None
        """

        templates.delete_template(template_id)


    def export_template(self, template_id, file_path=None):
        """ Export a template as XML, and optionally write it to a file or stdout.

        Arguments:
            template_id (str): ID of the template to export.
            file_path   (str): Optional, path of file to write the XML to.
        
        Returns:
            String (template xml)
        """

        template = templates.get_template(template_id, 'id').template

        output = 'file' if file_path else 'string'
        content = templates.export_template(template.id, output=output, file_path=file_path)

        return content 
    

    def import_template(self, file_path):
        """ Imports a template XML into Nifi's template store.

        Arguments:
            file_path   (str): Path of the XML file to import into Nifi as a template.
        
        Returns:
            None
        """

        canvas_id = canvas.get_root_pg_id()
        templates.upload_template(canvas_id, file_path)
        

nifi-deploy/nifi-deploy.py
import sys
import argparse

from nipyapi import templates, canvas
import nifi


def export_function(args):
    nifi_host       = args.nifi_host
    username        = args.username
    password        = args.password
    uuid            = args.uuid    
    name            = args.name
    description     = args.description
    filename        = args.filename
    keep_template   = args.keep_template

    try:
        n = nifi.NifiInstance(nifi_host, username=username, password=password)
    
        process_group = canvas.get_process_group(uuid, 'id')
        template = n.create_template(
            pg_id=uuid,
            name=name,
            desc=description
            ).template        

        xml = n.export_template(template.id, filename)
        if not filename and args.verbose: # Write to stdout
            print(xml)

    finally:
        if not args.keep_template:
            temp = templates.get_template_by_name(name)
            r = templates.delete_template(temp.id)
            if args.verbose:
                print(r.status_code, r.content.decode())


def import_function(args):
    nifi_host       = args.nifi_host
    username        = args.username
    password        = args.password
    filename        = args.filename

    n = nifi.NifiInstance(nifi_host, username=username, password=password)

    r = n.import_template(args.filename)
    if args.verbose:
        print(r.status_code, r.content.decode())


### argparse setup
parser = argparse.ArgumentParser(prog='nifi-deploy')
parser.add_argument('-v', '--verbose', action='store_true')
subparsers = parser.add_subparsers(title='Actions', help='Append --help to view action specific help')


example_usage = """EXAMPLE USAGE:
nifi-deploy export -n https://nifihost:9090 -u john -p badpractice 0a7361fd-015f-1000-ffff-ffffd2cbc7a7 my_great_template -d template description -f c:\\temp\\my_great_template_export.xml
"""
parser_export = subparsers.add_parser('export', epilog=example_usage)
parser_export.add_argument('uuid', help='UUID of a process group to export as a template XML')
parser_export.add_argument('name', help='Name of exported template - also the name internally if `--keep_template` is included')
parser_export.add_argument('-n', '--nifi_host', help='Nifi host (typically http://<nifihost>:8080), else using environment variable `NIFI_HOST`')
parser_export.add_argument('-u', '--username', help='Nifi username, else using environment variable `NIFI_USERNAME`')
parser_export.add_argument('-p', '--password', help='Nifi password, else using environment variable `NIFI_PASSWORD`')
parser_export.add_argument('-d', '--description', help='Description of the template - also used internally if `--keep_template` is included')
parser_export.add_argument('-f', '--filename', help='Path to save exported XML to, else outputs to stdout')
parser_export.add_argument('-k', '--keep_template', action='store_true', default=False, help='Keep the template in Nifi after export, else exclusively export XML and delete the temporarily instantiated template')
parser_export.set_defaults(func=export_function)


example_usage = """EXAMPLE USAGE:
nifi-deploy import -n https://nifihost:9090 -u john -p badpractice c:\\temp\\my_great_template_export.xml
"""
parser_import = subparsers.add_parser('import', epilog=example_usage)
parser_import.add_argument('filename', help='Path to template XML file for uploading to Nifi')
parser_import.add_argument('-n', '--nifi_host', help='Nifi host (typically http://<nifihost>:8080), else using environment variable `NIFI_HOST`')
parser_import.add_argument('-u', '--username', help='Nifi username, else using environment variable `NIFI_USERNAME`')
parser_import.add_argument('-p', '--password', help='Nifi password, else using environment variable `NIFI_PASSWORD`')
parser_import.set_defaults(func=import_function)


args = parser.parse_args()
if getattr(args, 'func', None):
    args.func(args)
else:
    parser.print_help()


nifi_deploy/nifi_deploy.py
""" 
This file is used for the CLI tool.
By default setup.py links `nifi-deploy` as an executable
to this file.
If you need to run this from source, run as a module 
with the `python -m` syntax.
"""

import argparse

from nifi_deploy.nifi import NifiInstance
from nifi_deploy import __version__


def export_function(args):
    nifi_host       = args.nifi_host
    username        = args.username
    password        = args.password
    uuid            = args.uuid    
    name            = args.name
    description     = args.description
    filename        = args.filename
    quiet           = args.quiet
    keep_template   = args.keep_template

    n = NifiInstance(nifi_host, username=username, password=password)
    template_id = None

    template = n.create_template(
        pg_id=uuid,
        name=name,
        desc=description
        )
    
    if not template:
        return
    
    content = n.export_template(template.id, filename)

    if not args.quiet:
        print(content)
    
    if not args.keep_template and template:
        n.delete_template(template.id)


def import_function(args):
    nifi_host       = args.nifi_host
    username        = args.username
    password        = args.password
    filename        = args.filename

    n = NifiInstance(nifi_host, username=username, password=password)

    n.import_template(args.filename)


def cli():
    ### argparse setup
    parser = argparse.ArgumentParser(prog='nifi-deploy')
    parser.add_argument('-v', '--version', action='store_true', default=False)
    subparsers = parser.add_subparsers(title='Actions', help='Append `--help` to view action specific help')


    ### export subparser
    example_usage = """EXAMPLE USAGE:
    nifi-deploy export -n https://nifihost:9090 -u john -p badpractice 0a7361fd-015f-1000-ffff-ffffd2cbc7a7 my_great_template -d template description -f c:\\temp\\my_great_template_export.xml --keep_template
    """
    parser_export = subparsers.add_parser('export', epilog=example_usage)
    parser_export.add_argument('uuid', help='UUID of a process group to export as a template XML')
    parser_export.add_argument('name', help='Name of exported template - also the name internally if `--keep_template` is included')
    parser_export.add_argument('-n', '--nifi_host', help='Nifi host (typically http://<nifihost>:8080), else using environment variable `NIFI_HOST`')
    parser_export.add_argument('-u', '--username', help='Nifi username, else using environment variable `NIFI_USERNAME`')
    parser_export.add_argument('-p', '--password', help='Nifi password, else using environment variable `NIFI_PASSWORD`')
    parser_export.add_argument('-d', '--description', help='Description of the template - also used internally if `--keep_template` is included')
    parser_export.add_argument('-f', '--filename', help='Path to save exported XML to')
    parser_export.add_argument('-q', '--quiet', action='store_true', default=False, help='Do not output exported XML to stdout')
    parser_export.add_argument('-k', '--keep_template', action='store_true', default=False, help='Keep the template in Nifi after export, else exclusively export XML and delete the temporarily instantiated template')
    parser_export.set_defaults(func=export_function)


    ### import subparser
    example_usage = """EXAMPLE USAGE:
    nifi-deploy import -n https://nifihost:9090 -u john -p badpractice c:\\temp\\my_great_template_export.xml
    """
    parser_import = subparsers.add_parser('import', epilog=example_usage)
    parser_import.add_argument('filename', help='Path to template XML file for uploading to Nifi')
    parser_import.add_argument('-n', '--nifi_host', help='Nifi host (typically http://<nifihost>:8080), else using environment variable `NIFI_HOST`')
    parser_import.add_argument('-u', '--username', help='Nifi username, else using environment variable `NIFI_USERNAME`')
    parser_import.add_argument('-p', '--password', help='Nifi password, else using environment variable `NIFI_PASSWORD`')
    parser_import.set_defaults(func=import_function)


    ### show usage help
    args = parser.parse_args()
    if getattr(args, 'version', None):
        print(__version__)
        return
    if getattr(args, 'func', None):
        args.func(args)
    else:
        parser.print_help()


if __name__ == '__main__':
    cli()